src/auth.test.js

```diff
@@ -0,0 +1,50 @@
+/* globals describe it jest expect */
+const { app } = require('../server')
+const request = require('supertest')
+const jwt = require('jsonwebtoken')
+
+// Prevent actual db writes/reads
+jest.mock('./lib/db', () => ({
+  getData: jest.fn().mockResolvedValue({}),
+  setData: jest.fn().mockReturnValue(null)
+}))
+
+describe('auth (js5/p6)', () => {
+  const mockUser = { username: 'tom', email: 'tom@tom' }
+
+  const validResponse = expect.objectContaining({
+    ...mockUser,
+    jwt: expect.any(String)
+  })
+
+  // set real jwt when creating user
+  let realJwt
+  it('creating a user return a jwt', async () => {
+    const response = await request(app)
+      .post('/auth/api/users')
+      .send({
+        ...mockUser,
+        password: 'anything'
+      })
+
+    expect(response.body).toEqual(validResponse)
+    realJwt = response.body.jwt
+  })
+  it('can get session with jwt', async () => {
+    const response = await request(app)
+      .get('/auth/api/session')
+      .set('Authorization', `Bearer ${realJwt}`)
```

Comment:
Usually a test shouldn't depend on the result of another test to make it work. This can cause issues if the tests are executed in a different order or are run in parallel. Test pre-conditions are usually set in the test or with an `beforeEach` block if the condition is needed for many tests.

Also, since we only need to test the app behavior to see if it handles all the cases as expected, I don't see a problem in mocking `jwt` here to make things easier. Otherwise you're just writing tests for `jwt` to see if it generates and verifies a token correctly, which is not really what we need, right?

Reply:
Ya I didn't like my tests depending on one another but I wasn't sure how to do the mocking and demonstrate the actual bug. Let me think about this for a bit and see what I can do.

Reply:
I'm just not seeing the easy solution to decouple the tests. The first test sets up the user and jwt that the next two count on. It was my understanding that same file tests always ran sync and it was only separate test files that ran async

Also if I mock jwt then it feels like I will miss the real bug which is the app using decode instead of verify.

Reply:
Just make sure to setup the pre-conditions the test needs inside it and don't rely on another test to do that. It's a good practice to have each test isolated from each other unless you have a good reason not to.

For a simple example of why we do that, imagine that suddenly we had a bug with user creation that caused an error when signing up. Since every test here assumes that a user will be created then all these test would fail, resulting in more work in finding the bug. Had every test be isolated, you'd quickly find that the bug was only happening when creating a user. That's why it's called *unit* tests, you have to test each unit of code separately.

In this case you only need to have a user created and a token assigned to it. That's all the pre-conditions to make the test work as you expect.

```js
const { saveUser } = require('./lib/userlist')
const { JWT_PRIVATE_SECRET } = require('./lib/fakeEnv')

describe('auth (js5/p6)', () => {

  // ...

  it('can get session with jwt', async () => {
    const token = jwt.sign({ username: mockUser.username }, JWT_PRIVATE_SECRET)
    saveUser({ ...mockUser, jwt: token })
    const response = await request(app)
      .get('/auth/api/session')
      .set('Authorization', `Bearer ${token}`)

    expect(response.body).toEqual(validResponse)
  })
  
  // ...
})
```

In cases like this with a real db we'd usually rollback the changes or clear the table so we can get a clean state for each test. But since this uses a simple object to keep the data we'd need to add another function to clear it inside a `beforeEach` or `afterEach` block.

Reply:
Thanks for the insight/suggestions. I pushed up the changes, I also included my attempt at cleaning up the tests were I just couldn't quite figure it out.